Objects (#321)
javascript-tutorial/es.javascript.info: 1-js/04-object-basics/01-object/article.md

# Objetos

Como aprendimos desde el capítulo <info:types>, hay ocho tipos de datos en JavaScript. Siete de ellos se denominan "primitivos", porque sus valores contienen solo un dato (sea un `string`, un número o lo que sea).

> En contraste, los objetos son usados para almacenar colecciones de varios datos y entidades más complejas asociados con un nombre clave. En JavaScript, los objetos penetran casi todos los aspectos del lenguaje. Por lo tanto, debemos comprenderlos primero antes de profundizar en cualquier otro lugar. llaves

Reviewer: ??? de donde habrá salido esto...

```suggestion
En contraste, los objetos son usados para almacenar colecciones de varios datos y entidades más complejas asociados con un nombre clave. En JavaScript, los objetos penetran casi todos los aspectos del lenguaje. Por lo tanto, debemos comprenderlos primero antes de profundizar en cualquier otro lugar. 
```

alert(user.name); // Pete
```

Puede parecer que la linea `(*)` ocasionaría un error, pero no. El `const` restablece el valor de `user` pero no su contenido.

El `const` podría dar error solo si intentamos asignar `user=...` en la totalidad de su contenido.

Hay otra manera de crear objetos con propiedades constantes, las veremos después en el capítulo <info:property-descriptors>.
````

> ## Llaves

Reviewer: Corregiste de más...  Acá sí son corchetes
corchetes  SQUARE  [  ]    para las propiedadess    objeto["nombre de propiedad"]

```suggestion
## Corchetes
```

Author: Jajajaja, lo hice a lo bestia y bien envalentonado les di cuello a todos los corchetes... Ya vi a qué te referías con el comentario de arriba. @joaquinelio

La notación de punto no funciona para acceder a propiedades con claves de más de una palabra:

```js run
// Esto nos daría un error de sintaxis
user.likes birds = true
```

JavaScript no entiende eso. Piensa que hemos accedido a `user.likes` y entonces nos da un error de sintaxis cuando aparece el inesperado `birds`.

El punto requiere que la clave sea un identificador de variable válido. Eso implica que: no contenga espacios, no comience con un dígito y no incluya caracteres especiales (`$` y `_` sí se permiten).

> Existe una "notación de llaves" alternativa que funciona con cualquier string:

Reviewer: ```suggestion
Existe una "notación de corchetes" alternativa que funciona con cualquier string:
```

```js run
let user = {};

// asignando
user["likes birds"] = true;

// obteniendo
alert(user["likes birds"]); // true

// eliminando
delete user["likes birds"];
```

> Ahora todo está bien. Nota que el string dentro de las llaves está adecuadamente entre comillas (cualquier tipo de comillas servirían).

Reviewer: ```suggestion
Ahora todo está bien. Nota que el string dentro de los corchetes está adecuadamente entre comillas (cualquier tipo de comillas servirían).
```

alert( user.key ) // undefined
```

### Propiedades calculadas

> Podemos usar llaves en un objeto literal al crear un objeto. A esto se le llama  *propiedades calculadas*.

Reviewer: ```suggestion
Podemos usar corchetes en un objeto literal al crear un objeto. A esto se le llama  *propiedades calculadas*.
```

Por ejemplo:

```js run
let fruit = prompt("¿Qué fruta comprar?", "Manzana");

let bag = {
*!*
  [fruit]: 5, // El nombre de la propiedad se obtiene de la variable fruit
*/!*
};

alert( bag.apple ); // 5 si fruit es="apple"
```

El significado de una propiedad calculada es simple: `[fruit]` significa que se debe tomar la clave de la propiedad `fruit`.

> Entonces, si un visitante ingresa `"apple"`, `bag` se convertira en `{apple: 5}`.

Reviewer: ```suggestion
Entonces, si un visitante ingresa `"apple"`, `bag` se convertirá en `{apple: 5}`.
```

Esencialmente esto funciona igual que:
```js run
let fruit = prompt("¿Qué fruta comprar?", "Manzana");
let bag = {};

// Toma el nombre de la propiedad de la variable fruit
bag[fruit] = 5;
```

...Pero luce mejor.

> Podemos usar expresiones más complejas dentro de las llaves:

Reviewer: ```suggestion
Podemos usar expresiones más complejas dentro de los corchetes:
```

};
```

> Las llaves son mucho más potentes que la notación de punto. Permiten cualquier nombre de propiedad y variables. Pero también son más engorrosos de escribir.

Reviewer: ```suggestion
Los corchetes son mucho más potentes que la notación de punto. Permiten cualquier nombre de propiedad y variables. Pero también son más engorrosos de escribir.
```

> Entonces, la mayoría de las veces, cuando los nombres de propiedad son conocidos y simples, se utiliza el punto. Y si necesitamos algo más complejo, entonces cambiamos a llaves.

Reviewer: ```suggestion
Entonces, la mayoría de las veces, cuando los nombres de propiedad son conocidos y simples, se utiliza el punto. Y si necesitamos algo más complejo, entonces cambiamos a corchetes.
```

alert( obj.for + obj.let + obj.return );  // 6
```

En resumen, no hay limitaciones en los nombres de propiedades. Pueden ser cadenas o símbolos (un tipo especial para identificadores que se cubrirán más adelante).

Otros tipos se convierten automáticamente en cadenas.

Por ejemplo, un número `0` se convierte en cadena `"0"` cuando se usa como clave de propiedad:

```js run
let obj = {
  0: "test" // igual que "0": "test"
};

// ambos alerts acceden a la misma propiedad (el número 0 se convierte a una cadena "0")
alert( obj["0"] ); // test
alert( obj[0] ); // test (la misma propiedad)
```

Hay una pequeña sorpresa por una propiedad especial llamada `__proto__`. No podemos establecerlo dentro de un valor que no sea de objeto:

```js run
let obj = {};
obj.__proto__ = 5; // asignando un número

> alert(obj.__proto__); // [objeto Object] - el valor es un objeto, no funciona como se "debería"

Reviewer: ```suggestion
alert(obj.__proto__); // [objeto Object] - el valor es un objeto, no funciona como "debería"
```
tr

```

Como podemos ver en el código, se ignora la asignación de un valor primitivo `5`.

Veremos la naturaleza especial de `__proto__` en los [capítulos siguientes](info:prototype-inheritance), y sugeriremos las [formas de arreglar](info:prototype-methods) tal comportamiento.

## La prueba de propiedad existente, el operador "in"

Una notable característica de los objetos en JavaScript, en comparación con muchos otros lenguajes, es que es posible acceder a cualquier propiedad. ¡No habrá error si la propiedad no existe!

La lectura de una propiedad no existente solo devuelve `undefined`. Así que podemos probar fácilmente si la propiedad existe:

```js run
let user = {};

alert( user.noSuchProperty === undefined ); // true significa que "no existe tal propiedad"
```

También existe un operador especial para ello: `"in"`.

La sintaxis es:
```js
"key" in object
```

Por ejemplo:

```js run
let user = { name: "John", age: 30 };

alert( "age" in user ); // mostrará "true", user.age sí existe

> alert( "blabla" in user ); // mostará false, user.blabla no existe

Reviewer: ```suggestion
alert( "blabla" in user ); // mostrará false, user.blabla no existe
```

```

> Nota que a la izquierda de `in` debe estar el *nombre de la propiedad* que suele ser un string citado.

Reviewer 2: ```suggestion
Nota que a la izquierda de `in` debe estar el *nombre de la propiedad* que suele ser un string entre comillas.
```

Si omitimos las comillas significa una variable. Esta variable debería almacenar el nombre real que será probado. Por ejemplo:

```js run
let user = { age: 30 };

let key = "age";

> alert( *!*key*/!* in user ); // true, porque su propiedad "age" si existe dentro del objeto

Reviewer: ```suggestion
alert( *!*key*/!* in user ); // true, porque su propiedad "age" sí existe dentro del objeto
```

```

Pero... ¿Por qué existe el operador `in`? ¿No es suficiente comparar con `undefined`?

La mayoría de las veces las comparaciones con `undefined` funcionan bien. Pero hay un caso especial donde esto falla y aún así `"in"` funciona correctamente.

Es cuando existe una propiedad de objeto, pero almacena  `undefined`:

```js run
let obj = {
  test: undefined
};

alert( obj.test ); // es undefined, entonces... ¿Quiere decir realmente existe tal propiedad?

alert( "test" in obj ); //es true, ¡La propiedad sí existe!
```

En el código anterior, la propiedad  `obj.test` técnicamente existe. Entonces el operador `in` funciona correctamente.

> Situaciones como esta suceden raramente ya que `undefined` no debe ser explícitamente asignado. Comunmente usamos `null` para valores "desconocidos" o "vacios". Por lo que el operdaor `in` es un invitado exótico en nuestro código.

Reviewer: ```suggestion
Situaciones como esta suceden raramente ya que `undefined` no debe ser explícitamente asignado. Comúnmente usamos `null` para valores "desconocidos" o "vacíos". Por lo que el operador `in` es un invitado exótico en nuestro código.
```
Me suena horrible, pero la norma dice que el acento se arrastra y se transforma en esdrújula
me preguntaste por fundeu...
www.fundeu.es/consulta/comun-comunmente-2784/

## El bucle "for..in" 

Para recorrer todas las claves de un objeto existe una forma especial de bucle: `for..in`. Esto es algo completamente diferente a la construcción `for(;;)` que estudiaremos más adelante.

La sintaxis:

```js
for (key in object) {

>   // se ejecuta e cuerpo para cada clave entre las propiedades del objeto

Reviewer: ```suggestion
  // se ejecuta el cuerpo para cada clave entre las propiedades del objeto
```

};

for (let key in user) {
  // claves
  alert( key );  // name, age, isAdmin
  // valores de las claves
  alert( user[key] ); // John, 30, true
}
```

Nota que todas las construcciones "for" nos permiten declarar variables para bucle dentro del bucle, como `let key` aquí.

> Además podriamos usar otros nombres de variables en lugar de `key`. Por ejemplo, `"for (let prop in obj)"` también se usa bastante.

Reviewer: ```suggestion
Además podríamos usar otros nombres de variables en lugar de `key`. Por ejemplo, `"for (let prop in obj)"` también se usa bastante.
```

### Ordenado como un objeto

> ¿Los objetos son ordenados? Es decir, si creamos un bucle sobre un objeto, ¿obtenemos todas las propiedades en el mismo orden en el que se agregaron? ¿Podemos confiar en ello?

Reviewer 2: Me parece que `están` es más adecuado pues se está hablando del estado del objeto.
```suggestion
¿Los objetos están ordenados? Es decir, si creamos un bucle sobre un objeto, ¿obtenemos todas las propiedades en el mismo orden en el que se agregaron? ¿Podemos confiar en ello?
```

> La respuesta corta es: "ordenados de una forma especial": las propiedades de enteros se ordenan, los demás aparecen en el orden de la creación. Entremos en detalle.

Reviewer 2: Me parece que enteros necesita números para ser más claro. O conservar `integer`
```suggestion
La respuesta corta es: "ordenados de una forma especial": las propiedades de números enteros se ordenan, los demás aparecen en el orden de la creación. Entremos en detalle.
```

*/!*
```

El objeto puede usarse para sugerir una lista de opciones al usuario. Si estamos haciendo un sitio principalmente para el público alemán, entonces probablemente queremos que `49` sea el primero.

Pero si ejecutamos el código, veremos una imagen totalmente diferente:

- USA (1) va primero
- Luego Switzerland (41) y así sucecivamente.

Los códigos telefónicos van en orden ascendente porque son enteros. Entonces vemos  `1, 41, 44, 49`.

> ````smart header="¿Propiedades de enteros? ¿Qué es eso?"
> El término "propiedad de enteros" aquí significa que una cadena se puede convertir a y desde desde un entero sin nigún cambio.

Reviewer 2: ```suggestion
````smart header="¿Propiedades de números enteros? ¿Qué es eso?"
El término "propiedad de enteros" aquí significa que una cadena se puede convertir a y desde desde un número entero sin nigún cambio.
```

Entonces, "49" es un nombre de propiedad entero, porque cuando este se transforma a un entero y viceversa continua siendo el mismo. Pero "+49" y "1.2" no lo son:

```js run

> // Math.trunc es una función incorporada que elimina la parte decimal

Reviewer 2: `built-in` se refiere a las funciones que vienen por defecto dentro de JavaScript
```suggestion
// Math.trunc es una función nativa que elimina la parte decimal
```

}
```

Ahora sí funciona como debería.

## Resumen

Los objetos son arreglos asociativos con varias características especiales.

Almacenan propiedades (pares de clave-valor), donde:
- Las claves de propiedad deben ser cadenas o símbolos (generalmente strings).
- Los valores pueden ser de cualquier tipo.

Para acceder a una propiedad, podemos usar:
- La notación de punto: `obj.property`.
- Notación de corchetes `obj["property"]`. Los corchetes permiten tomar la clave de una variable, como `obj[varWithKey]`.

Operadores adicionales:
- Para eliminar una propiedad: `delete obj.prop`.

> -Para comprobar si existe una propiedad con la clave proporcionada: `"key" in obj`.

Reviewer 2: ```suggestion
- Para comprobar si existe una propiedad con la clave proporcionada: `"key" in obj`.
```

> - Para crear bluces sobre un objeto: bucle `for (let key in obj)`.

Reviewer: ```suggestion
- Para iterar sobre un objeto: bucle `for (let key in obj)`.
```
iteracion ya la usamos en varios lados

Lo que hemos estudiado en este capítulo se llama "objeto simple", o solamente `Object`.

Hay muchos otros tipos de objetos en JavaScript:

- `Array` para almacenar colecciones de datos ordenados,
- `Date` para almacenar la información sobre fecha y hora,
- `Error` para almacenar información sobre un error.
- ...Y así.

> Tienen sus características especiales que estudiaremos más adelante. A veces las personas dicen algo como "Tipo de matriz" o "Tipo de fecha", pero formalmente no son tipos en sí, sino que pertenecen a un tipo de datos de "objeto" simple y lo amplian a varias maneras.

Reviewer 2: Me parece que no se deberían traducir `Array` y `Date` pues hacen referencia a tipos de estructuras de datos del lenguaje.
```suggestion
Tienen sus características especiales que estudiaremos más adelante. A veces las personas dicen algo como "Tipo Array" o "Tipo Date", pero formalmente no son tipos en sí, sino que pertenecen a un tipo de datos de "objeto" simple y lo amplian a varias maneras.
```
